pkgs/applications/science/chemistry/cp2k/default.nix

```diff
@@ -64,28 +117,58 @@ in stdenv.mkDerivation rec {
     FC         = mpif90
     LD         = mpif90
     AR         = ar -r
+    ${lib.strings.optionalString (gpuBackend == "cuda") ''
+    OFFLOAD_CC = nvcc
+    OFFLOAD_FLAGS = -O3 -g -w --std=c++11
+    OFFLOAD_TARGET = cuda
+    GPUVER = ${gpuVersion}
+    CXX = mpicxx
+    CXXFLAGS = -std=c++11 -fopenmp
+    ''}
+    ${lib.strings.optionalString (gpuBackend == "rocm") ''
+    GPUVER = ${gpuVersion}
+    OFFLOAD_CC = hipcc
+    OFFLOAD_FLAGS = -fopenmp -m64 -pthread -fPIC -D__GRID_HIP -O2 --offload-arch=${gpuArch} --rocm-path=${rocm-core}
+    OFFLOAD_TARGET = hip
+    CXX = mpicxx
+    CXXFLAGS = -std=c++11 -fopenmp -D__HIP_PLATFORM_AMD__
+    ''}
     DFLAGS     = -D__FFTW3 -D__LIBXC -D__LIBINT -D__parallel -D__SCALAPACK \
                  -D__MPI_VERSION=3 -D__F2008 -D__LIBXSMM -D__SPGLIB \
                  -D__MAX_CONTR=4 -D__LIBVORI ${lib.optionalString enableElpa "-D__ELPA"} \
-                 -D__PLUMED2
-    CFLAGS    = -fopenmp
+                 -D__PLUMED2 -D__HDF5 -D__GSL -D__SIRIUS -D__LIBVDWXC -D__SPFFT -D__SPLA \
+                 ${lib.strings.optionalString (gpuBackend == "cuda") "-D__OFFLOAD_CUDA -D__DBCSR_ACC"} \
+                 ${lib.strings.optionalString (gpuBackend == "rocm") "-D__OFFLOAD_HIP -D__DBCSR_ACC -D__NO_OFFLOAD_PW"}
+    CFLAGS    = -fopenmp -I${hdf5-fortran}/include -I${gsl}/include
     FCFLAGS    = \$(DFLAGS) -O2 -ffree-form -ffree-line-length-none \
                  -ftree-vectorize -funroll-loops -msse2 \
                  -std=f2008 \
                  -fopenmp -ftree-vectorize -funroll-loops \
+                 -I${libint}/include ${lib.optionalString enableElpa "$(pkg-config --variable=fcflags elpa)"} \
+                 -I${sirius}/include/sirius \
                  -I${lib.getDev libxc}/include -I${lib.getDev libxsmm}/include \
-                 -I${libint}/include ${lib.optionalString enableElpa "$(pkg-config --variable=fcflags elpa)"}
+                 ${lib.strings.optionalString enableSSE41 "-msse4_1"} \
```

Comment:
I am not sure that is a good idea introducing the optimization flags here. We could do the same thing for almost every package, which would introduce quite a lot of boiler plate code.

In NixQChem, these flags (or rather the march/mtune which imply these flags) are injected into the optimized stdenv.

Reply:
Hm I know and I was thinking about this, too. However, I am in favour of writing this boiler plate here. The mechanism to do so correctly is directly supported by Nixpkgs and used in a handful, altough not an enormous amount, of places. Furthermore I am in favour having a performant version of stuff we've upstreamed in also in Nixpkgs. It shouldn't be necessary to utilise NixOS-QChem for a package that is upstream to get good performance. The last aspect is, that the override of stdenv feels somewhat hacky too me. Currently, with `config.qchem.optArch` we are basically replicating what `localSystem.gcc.arch` and `localSystem.gcc.tune` are doing. I would be in favour of using this mechanism instead of `config.qchem.optArch` and use optimisation flags where available instead of the stdenv override.

Reply:
> The mechanism to do so correctly is directly supported by Nixpkgs and used in a handful, although not an enormous amount, of places.

The places where I have used these flags, I have only used them set configure/cmake flags, but not to directly set gcc flags.

>  It shouldn't be necessary to utilise NixOS-QChem for a package that is upstream to get good performance.

Agreed, but I think this is the wrong place here. This should be fixed on another level, so that it becomes a more general valid solution. Adding this many gcc optimization flags to every package is not sustainable. I would remove them here and rather start the discussion of how to solve the problem properly on a nixpkgs wide basis.

> The last aspect is, that the override of stdenv feels somewhat hacky too me.

It actually is not even that hacky. Moreover, this is the purpose of NixOS-QChem, to implement solutions quickly, that are not fully up nixpkgs standard.

> Currently, with config.qchem.optArch we are basically replicating what localSystem.gcc.arch and localSystem.gcc.tune are doing.

There is actually fine difference here: if you use set localSystem.gcc.arch, you get a complete mass rebuild, which even rebuilds stdenv with the new flags. The last time I tried this, it failed to build (you would need to fix a bunch of packages that are dependencies but not very relevant.). If I remember correctly, you then also have built a new architecture (based on what arch you set) you are effectively now cross compiling.
The config.qchem.optArch solution is more granular, it only injects the gcc flags into the environment but does not require a rebuild of stdenv.

Reply:
> Agreed, but I think this is the wrong place here. This should be fixed on another level, so that it becomes a more general valid solution. Adding this many gcc optimization flags to every package is not sustainable. I would remove them here and rather start the discussion of how to solve the problem properly on a nixpkgs wide basis.

Hm, I've no other idea where one would start such a discussion? On the discourse? I'm not sure if I understand the entirety of the problem. Basically cross-compiling sounds like a very good idea to me here and I'd rather see the cross compilation failures fixed. I see that the stdenv injection is more granular, but not cross-compiling seems more restrictive to me. Like failures if the build machine does not have AVX2 for example, no?

> The places where I have used these flags, I have only used them set configure/cmake flags, but not to directly set gcc flags.

Hm, fair enough. But CP2K has no configure and cmake, so there is no other mechanism than setting them directly?

> Adding this many gcc optimization flags to every package is not sustainable.

At least the HPC stuff, no? Even quite basic stuff such as Qt sets these flags. 
https://github.com/NixOS/nixpkgs/blob/c2de65438c7306cc6b8214d805ef657b8b5fcfa9/pkgs/development/libraries/qt-5/modules/qtbase.nix#L252 or openssl 
https://github.com/NixOS/nixpkgs/blob/3fd75f93ab155e6d9467e19d31d777b76ccd3183/pkgs/development/libraries/openssl/default.nix#L6
or some biology packages https://github.com/NixOS/nixpkgs/blob/c2de65438c7306cc6b8214d805ef657b8b5fcfa9/pkgs/applications/science/biology/kalign/default.nix#L8

All these flags are handled more or less automatically via the stdenv mechanism?

> It actually is not even that hacky. Moreover, this is the purpose of NixOS-QChem, to implement solutions quickly, that are not fully up nixpkgs standard.

Fair enough :slightly_smiling_face: :smile: 
I've removed them from CP2K for now

> The config.qchem.optArch solution is more granular, it only injects the gcc flags into the environment but does not require a rebuild of stdenv.

Yes, I see this point. But this is also possible with enabling the AVX flag manually without overriding the gcc arch on all packages.